---
title: "Figures for the Paper"
author: "Aina Vaquer Pic√≥"
date: "2025-05-02"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

```

##  Loading necessary libraries

```{r  warning=FALSE, message=FALSE}

library(tidyr)
library(dplyr)
library(ggplot2)
library(readr)
library(patchwork)
library(gtools)
library(gridExtra)
library(ggplotify)
library(stringr)

```

### tDNAs accumulate a higher number of mutations than other genes in the genome

Figure 1A

```{r fig.width=14, fig.height = 6, fig.align='center', warning=FALSE}

# tRNA genes
new = read_csv("/data/gtl/avaquer/tRNA_mut_density/data/sum_count_density_with_trint.csv",  show_col_types = FALSE)


#new = read_csv( "/g/strcombio/fsupek_franklin/cancer3_backup/mmurillo/results/mut_sum/sum_count_density_cancer_samples.csv")

#Normalized density - Normalization each group (expression level) should be normalized separately by the total number of sum_density in that group; otherwise, they are not comparable.
new$group1 = paste0(new$level)
new1 = dplyr::bind_rows(lapply(split(new, new$group1), function(d){
  d$norm_density = d$sum_density/(sum(d$sum_density))
  return(d)
}))

# all windows by expression
custom_colors <- c('#ffc6da','#ff7ead','#cc4778','#973458')
figure1A <- ggplot(new1, aes(x = as.numeric(i), y = norm_density, col = level)) +
  geom_line(linewidth=1.4) + geom_point(size=2.2) +
  geom_vline(xintercept = 0, linetype = "dashed", color = 'black',linewidth = 1) + facet_wrap(vars(level), nrow = 1) +
  scale_color_manual(values = custom_colors) +
  theme_classic() +
  theme(text = element_text(size = 14),  
    axis.text = element_text(size = 14), 
    axis.title = element_text(size = 16),  
    strip.background = element_blank(),
    strip.text = element_text(size = 14),
    legend.position = "none") +
  xlim(-25,25) + 
  labs(x = "Distance to tRNA genes (x100nts)", y = "Normalized\nmutation density") 

### Protein coding genes

new = read_csv( "/data/gds/franklin/cancer3_backup/mmurillo/results/mut_rates_per_proteinCoding/sum_count_density_proteincoding_by_expression.csv", show_col_types = FALSE)

#Normalization
new$group1 = paste0(new$level)
new1 = dplyr::bind_rows(lapply(split(new, new$group1), function(d){
  d$norm_density = d$sum_density/(sum(d$sum_density))
  return(d)
}))

max_prot <- max(new1$norm_density[new1$i == 0])

muts_in_protein <- new1[new1$i == 0,]


# All windows and plot by expression separated  

custom_colors <- c('#fed1ad','#feb980','#f89540','#d96705')
sum_count_density_with_trint <- ggplot(new1, aes(x = as.numeric(i), y = norm_density, col = level)) +
  geom_line(linewidth=1.4) + geom_point(size=2.2) +
  geom_vline(xintercept = 0, linetype = "dashed", color = 'black',linewidth = 1) + facet_wrap(vars(level), nrow = 1) +
  scale_color_manual(values = custom_colors) +
  theme_classic() +
  theme(text = element_text(size = 14),  
    axis.text = element_text(size = 14), 
    axis.title = element_text(size = 16),  
    strip.text = element_text(size = 14),
    strip.background = element_blank(),
    legend.position = "none") +
  xlim(-25,25) + 
  labs(x = "Distance to Protein-coding genes (x100nts)", y = "Normalized\nmutation density")
label_plot <- ggplot() + 
  theme_void() + 
  annotate("label", x = 1, y = 0.5, label = "Protein-Coding\nGenes",
           angle = 270,  # rotate label for vertical text
           fill = "grey90", color = "black", 
           label.size = 1, size = 6,
           label.r = unit(0.2, "lines"))

# Combine with patchwork
figure1B  <- sum_count_density_with_trint + label_plot + plot_layout(widths = c(1, 0.1))



```

Figure 1B

```{r fig.width=14, fig.height = 12, fig.align='center', warning=FALSE}

rRNAs <- read.csv("/data/gtl/avaquer/tRNA_mut_density/data/sum_count_density_cancer_samples_rRNAs.csv")
snRNAs <- read.csv("/data/gtl/avaquer/tRNA_mut_density/data/sum_count_density_cancer_samples_snRNAs.csv")
misc_RNAs <- read.csv("/data/gtl/avaquer/tRNA_mut_density/data/sum_count_density_cancer_samples_misc_RNAs.csv")
rRNAs$type <- rep("rRNA", nrow(rRNAs))
snRNAs$type <- rep("snRNA", nrow(snRNAs))
misc_RNAs$type <- rep("miscRNA", nrow(misc_RNAs))
plot_data <-  rbind(rRNAs, snRNAs, misc_RNAs)
plot_data <-  dplyr::bind_rows(lapply(split(plot_data, plot_data$type), function(d){
  d$norm_density = d$sum_density/(sum(d$sum_density))
  return(d)
}))
max_other <- max(plot_data$norm_density[plot_data$i == 0])
plot_data <- plot_data[plot_data$i >= -50 & plot_data$i <= 50, ]
figure1C <- ggplot(plot_data, aes(x = as.numeric(i), y = norm_density, col = type)) +
  geom_line(linewidth=1.1) +
  geom_point(size=1.2) +
  geom_vline(xintercept = 0, linetype = "dashed", color = 'black', linewidth = .8) +
  facet_wrap(vars(type)) +  # Faceting by type
  theme_classic()+
  xlim(-25,25) + 
  theme(
    text = element_text(size = 14),  
    axis.text = element_text(size = 14), 
    axis.title = element_text(size = 16),  
    strip.text = element_text(size = 14),
    strip.background = element_blank(),
    legend.position = "none") +
  labs(
    x = "Distance to other Pol3-Transcribed genes (x100nts)",
    y = "Normalized mutation\ndensity"
  ) + scale_color_manual(values = c("#f0f921", "#7e03a8", "#0d0887")) 

# Render the plot
label_plot <- ggplot() + 
  theme_void() + 
  annotate("label", x = 1, y = 0.5, label = "Other Pol3-Transcribed\nGenes",
         angle = 270,  # rotate label for vertical text
         fill = "grey90", color = "black", 
         label.size = 1, size = 6,
         label.r = unit(0.2, "lines"))


# Combine with patchwork
figure1C  <- figure1C + label_plot + plot_layout(widths = c(1, 0.1))


figure1A <- figure1A +
  geom_hline(yintercept = max_prot, color = "#d96705", linetype = "dashed", size = 1) +
  geom_hline(yintercept = max_other, color = "#0d0887", linetype = "dashed", size = 1)


 label_plot <- ggplot() + 
  theme_void() + 
  annotate("label", x = 1, y = 0.5, label = "tRNA Genes",
           angle = 270,  # rotate label for vertical text
           fill = "grey90", color = "black", 
           label.size = 1, size = 6,
           label.r = unit(0.2, "lines"))

# Combine with patchwork
figure1A  <- figure1A  + label_plot + plot_layout(widths = c(1, 0.1))

figure1 <- figure1A / figure1B / figure1C 

figure1


```

### Tumor-type dependence of tDNA somatic mutagenic rates

Figure 2

```{r fig.width=14, fig.height = 6, fig.align='center', warning=FALSE}

custom_colors <- c('#f0f921','#ed7953','#9c179e','#0d0887')

# log FC per cancer type
#new = read_csv("/g/strcombio/fsupek_home/mmurillo/results/sum_count_density_with_trint_byct.csv")
new = read_csv("/data/gds/franklin/cancer3_backup/mmurillo/results/mut_sum/sum_count_density_with_trint_byct_OK.csv",  show_col_types = FALSE)


b <- new[new$i == 0,]
new$group1 = paste0(new$level, "_", new$tissue)

new1 = dplyr::bind_rows(lapply(split(new, new$group1), function(d){

  res = unlist(lapply(seq(10, 90, 10), function(i){
    a = d$sum_density[d$i == 0] / mean(d$sum_density[d$i %in% i:(i+9)])
    b = d$sum_density[d$i == 0] / mean(d$sum_density[d$i %in% -i:(-i-9)])
    return(c(a,b))
  }))
  
  df = data.frame(tissue = unique(d$tissue), level = unique(d$level), 
                  age = unique(d$age_label), mean = mean(res), 
                  sd = sd(res), stringsAsFactors = F)
  return(df)
}))

new1[is.na(new1)] <- 0

red = new1[new1$level == "Level4",]
red = red[order(red$mean, decreasing = T),]
new1$tissue = factor(new1$tissue, levels = unique(red$tissue))
new1$log_mean = log(new1$mean)
new1 <- new1[!new1$tissue %in% c('unknown', 'thyroid'), ]
#new1 <- new1[!new1$tissue %in% c('thyroid'), ]

red$tissue <- str_replace_all(red$tissue, "_", " ")              
red$tissue <- str_replace(red$tissue, "^(\\w)", str_to_upper)
new1$tissue <- str_replace_all(new1$tissue, "_", " ")              
new1$tissue <- str_replace(new1$tissue, "^(\\w)", str_to_upper)
new1$tissue = factor(new1$tissue, levels = unique(red$tissue))



# Plot
trna_FC_cancer <- ggplot(new1, aes(x = tissue, y = mean, col = level)) +
  geom_hline(yintercept = 1, linetype = "dashed") +
  geom_point(size = 4, position=position_dodge(width = 0.5)) +
  geom_errorbar(aes(ymin=mean-sd, ymax=mean+sd), width=.0, position=position_dodge(width = 0.5)) +
  scale_color_manual(values = custom_colors) +
  scale_fill_manual(values = custom_colors) +
  ylab("FC in tRNA versus flanking regions") +
  xlab("Tissue")+
  labs(color = "Expression Level") +
  theme_classic()+ 
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        plot.title = element_text(hjust = 0.5, size = 18,margin = margin(b =20)),
        legend.position = c(0.94, 0.81),
        axis.text = element_text(size = 14),
        axis.title = element_text(size = 16),
        legend.text = element_text(size = 14),
        legend.title = element_text(size=16)) 



trna_FC_cancer

```

### Distribution of somatic mutations in tDNAs and evidence for negative selection

```{r}

# Calculate Mutational Density of all Positions in tRNAs

files = list.files("/data/gds/franklin/cancer3_backup/mmurillo/results/mut_density_per_tRNA_control_trint_PERPOS/", pattern = "*", full.names = T)
files = files[!grepl("NA.csv", files)]
df_files = data.frame(file_name = files)

combined_data <- bind_rows(lapply(files, function(file) {read.csv(file, header = TRUE, colClasses = c("character", "character", "numeric"))}))

positions_interest <- c(1:20, "20A", "20B", 21:73)
modified_positions <- c(4,6,7,9,10,12, 13, 14, 16, 17, 18, 20, "20A", "20B", 26:28, 31,32,34,35,37:40, 44,46:50, 54,55,58, 67, 72 )
conserved_positions <- c(8,14,18,19,21,33,53,54,55,56,61)
significant <- as.character(c(8,33,54,55,40, 56, 57, 68))
mut_density <- combined_data[combined_data$i %in% positions_interest, ] %>%
  group_by(i) %>%
  summarise(mean_density = mean(sum_density)) %>%
  mutate(significant = ifelse(i %in% significant, "Yes", "No")) %>%
  mutate(modified = ifelse(i %in% modified_positions, "Yes", "No")) %>%
  mutate(conserved = ifelse(i %in% conserved_positions, "Yes", "No"))

motifs <- read_table("../data/positions_def.tsv", show_col_types = FALSE)
colnames(motifs) <- c("i", "count", "TCN_percentage")


mut_density <- merge(mut_density, motifs[,c("i", "TCN_percentage")], by = "i")


```
Figure 3

```{r fig.width=14, fig.height = 7, fig.align='center'}

q05 <- quantile(mut_density$mean_density, probs = 0.05, na.rm = TRUE)
q95 <- quantile(mut_density$mean_density, probs = 0.95, na.rm = TRUE)
sorted_i <- mixedsort(unique(mut_density$i))
mut_density$i <- factor(mut_density$i, levels = sorted_i)
mut_density$significant <- factor(mut_density$significant, levels = c("No", "Yes"))
mut_density$modified <- factor(mut_density$modified, levels = c("No", "Yes"))

#TCN percentage

barplot_plot <- ggplot(mut_density, aes(x = i, y = TCN_percentage)) +
  geom_bar(stat = "identity", fill = "#5d5d5d", color = "#5d5d5d", alpha = 0.7) +
  labs(y = "% of APOBEC\nmotif TCN" ) +
  theme_classic() + 
  theme(
    axis.text.x = element_blank(),  # Remove x-axis labels
    axis.title.x = element_blank(), # Remove x-axis title
    axis.text.y = element_text(size = 14),
    plot.title = element_text(size = 18),
    axis.title.y = element_text(size = 16),  
    axis.ticks.x = element_blank()
  ) + 
  scale_y_continuous(breaks = seq(0, 0.75, by=0.25),
                     expand = c(0, 0))  # Remove gap between bars and x-axis 


# Mean density

zones <- data.frame(
  zone = factor(c("A-box", "Anticodon", "B-box"),
                levels = c("A-box", "Anticodon", "B-box")),
  start = c(8, 36, 54),
  end = c(20, 38, 64)
)
main_plot <- ggplot(mut_density, aes(x = i, y = mean_density, group = 1)) +
  # Don't map fill inside aes() for geom_rect
  geom_rect(data = zones, 
            inherit.aes = FALSE,
            aes(xmin = start - 0.5, xmax = end + 0.5, ymin = -Inf, ymax = Inf),
            fill = c('#F3D9E7','#E6F0F0', '#E6D9F0'), alpha = 1) +

  geom_hline(yintercept = q05, linetype = "dashed", color = "#0072B2", linewidth = 0.6) +
  geom_hline(yintercept = q95, linetype = "dashed", color = "#0072B2", linewidth = 0.6) +
  geom_line(color = "#878787", linewidth = 1) +
  geom_point(aes(fill = significant, color = modified, shape = conserved), size = 3, stroke = 1.2) +
  scale_fill_manual(values = c("Yes" = "#0072B2", "No" = "#878787")) +
  scale_color_manual(values = c("Yes" = "#E69F00", "No" = "#878787")) +
  scale_shape_manual(values = c("No" = 21, "Yes" = 22)) + 
  labs(
    x = "Position", 
    y = "Mean Mutational\nDensity"
  ) +
  theme_classic() + 
  theme(
    axis.text.x = element_text(angle = 90, hjust = 1, size = 14),
    axis.text.y = element_text(size = 14),
    plot.title = element_text(size = 18),
    legend.text = element_text(size = 14),
    axis.title.x = element_text(size = 16),
    axis.title.y = element_text(size = 16)
   
  )

combined_plot <- barplot_plot / main_plot
combined_plot



```

### tRNA mutagenesis and generation of chimeric tRNAs

Figure Supplementary

```{r}

files <- list.files("/data/gtl/avaquer/tRNA_mut_density/MutDensity/results_mut_density_chim_tRNA/", pattern = "*", full.names = T)
chim_density <- do.call(rbind, lapply(files, read.csv))
chim_density$tRNA <-  sub("-[0-9]+-[0-9]+$", "", chim_density$tRNA)
chim_density <- chim_density[chim_density$pos %in% c("35","36"),] %>% group_by(tRNA,pos,tissue) %>% summarise(mean_density  = mean(sum_density))
tissues_keep <- unique(chim_density$tissue[!chim_density$mean_density == 0])
chim_density <- chim_density[chim_density$tissue %in% tissues_keep,]
# Get all combinations of tRNA and tissue from the filtered dataset
tissue_trna_combos <- chim_density %>%
  distinct(tRNA, tissue)

# Create a data frame of all missing (or potentially missing) position 34 rows
pos34_df <- tissue_trna_combos %>%
  mutate(pos = 34, mean_density = 0)

# Bind this to the original dataset and remove duplicates if any
chim_density <- bind_rows(chim_density, pos34_df) %>%
  distinct(tRNA, tissue, pos, .keep_all = TRUE)

```

```{r fig.height=8, fig.width=12, fig.align='center'}

max_density <- max(chim_density$mean_density, na.rm = TRUE)

ggplot(chim_density, aes(x = pos, y = tRNA, fill = mean_density)) +
  geom_tile(color = "#908c8d") +
  scale_fill_gradientn(
    colors = c("gray90", "#f0bcd0", "#900C3F"),
    values = scales::rescale(c(0, 0.01 * max_density, max_density)),
    name = "Mutational Density"
  ) +
  facet_wrap(~ tissue, scales = "free_y") +
  labs(x = "Position", y = "tRNA") +
  theme_classic() +
  theme(
    axis.text.y = element_text(size = 14),
    axis.text.x = element_text(size = 14, angle = 45, hjust = 1),
    axis.title = element_text(size = 16),
    strip.text = element_text(size = 16),
    legend.title = element_text(size = 16),
    legend.text = element_text(size=14),
    strip.background = element_blank(),
    strip.placement = "outside",
    # Hide x-axis text and ticks for all but bottom panels
    panel.spacing = unit(1, "lines"),
    axis.text.x.top = element_blank(),
    axis.ticks.x.top = element_blank()
    
  )


```

Figure 5A

```{r fig.align='center', fig.width=6, fig.height=3}

files <- list.files("/data/gtl/avaquer/tRNA_mut_density/MutDensity/results_mut_density_chim_tRNA_all/", pattern = "*", full.names = T)
chim_density_all <- do.call(rbind, lapply(files, read.csv))
chim_density_all$tRNA <-  sub("-[0-9]+-[0-9]+$", "", chim_density_all$tRNA)
chim_density_all <- chim_density_all %>% group_by(tRNA,pos) %>% summarise(mean_density  = mean(sum_density), pos = as.character(pos))
max_density <- max(chim_density_all$mean_density, na.rm = TRUE)


ggplot(chim_density_all, aes(x = pos, y = tRNA, fill = mean_density)) +
  geom_tile(color = "#908c8d", linewidth = 1) +
  scale_fill_gradientn(
    colors = c("gray90", "#f0bcd0", "#900C3F"),
    values = scales::rescale(c(0, 0.01 * max_density, max_density)),
    name = "Mutational Density"
  ) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  labs(x = "Position", y = "tRNA") +
  theme_classic() +
  theme(
    axis.text.y = element_text(size = 14),
    axis.text.x = element_text(size = 14, hjust = 1),
    axis.title = element_text(size = 16),
    legend.title = element_text(size = 16),
    legend.text = element_text(size = 14),
    axis.line = element_blank(),     # removes axis lines
    axis.ticks = element_blank()
  )




```


Figure 5B

```{r fig.align='center', fig.width=10, fig.height=10}

data <- read.csv("../data/chimeric_trnas.txt", sep = "\t")
colnames(data)[length(colnames(data))] <- "tRNA_id"
data$tRNA_id <- paste(data$prefix, data$amino_acid, sep = "-")
plot_list <- list()
# Possible observed 
bases <- c("T", "G", "C", "A")
position <- c("34", "35", "36")
colors <- c("gray90","#fce5ee", "#f0bcd0","#c81057", "#900C3F")


for(tRNA in unique(data$tRNA_id)){
  plot <- data.frame(positions = character(),anticodon_base = character(), changed_base = character(), count = integer(), substituted_aa = character(), anticodon = character(), stringsAsFactors = FALSE)
  temp <- data[data$tRNA_id == tRNA,]
  for(anticodon in unique(temp$anticodon)){
    temp2 <- temp[temp$anticodon == anticodon,]
    base_vec <- strsplit(anticodon, "")[[1]]
      for (i in 1:3){
        # For each position, expand to all possible bases
        expanded_data <- data.frame(
          position = position[i],
          anticodon_base = base_vec[i],
          changed_base = bases,
          count = rep(0,4),
          substituted_aa = rep("aa",4),
          anticodon = rep(anticodon,4)
        )
        plot <- rbind(plot,expanded_data)
      }
    }
  for(num in 1:nrow(temp)){
    plot$count[plot$anticodon == temp$anticodon[num] & plot$anticodon_base == temp$ref[num] & plot$changed_base == temp$alt[num]] <- plot$count[plot$anticodon == temp$anticodon[num]  & plot$anticodon_base == temp$ref[num] & plot$changed_base == temp$alt[num]] + 1
  }
  plot$x_axis <- paste(plot$position, plot$anticodon_base, sep = "\n")
  plot$y_axis <- paste(plot$substituted_aa, plot$changed_base, sep = "\t")
  plot_section <- ggplot(plot, aes(x = x_axis, y = y_axis, fill = factor(count))) +
  geom_tile(color = "#908c8d", linewidth = 1) +
  facet_wrap(~ anticodon, strip.position = "top", scales = "free_x", ncol = 4) +
  scale_fill_manual(values = colors, name = "Count") +
  labs(
    title = tRNA,
    x = "",
    y = "Observed Nucleotide Change"
  ) +
  theme_classic() +
  theme(
    plot.title = element_text(size = 16),
    strip.placement = "outside",
    strip.text = element_text(size = 14),
    axis.text = element_text(size = 14),
    axis.ticks = element_blank(),
    axis.line = element_blank(),
    axis.title = element_text(size = 16, margin = margin(r = 20)),
    panel.spacing = unit(1, "lines"),
    legend.text = element_text(size=14),
    legend.title = element_text(size=16),
    strip.background = element_blank()
  )
  plot_list[[tRNA]] <- plot_section
}

grid.arrange(
  as.grob(plot_list[[3]]),
  as.grob(plot_list[[1]]),
  as.grob(plot_list[[2]]),
  as.grob(plot_list[[4]]),
  ncol = 1
)


```


